### PR TITLE
[processing] port XYZ tiles algorithms

### PR DESCRIPTION
**Pull request by alexbruy**
## Description

Port XYZ tiling algorithms to C++ as a part of [QEP #271](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/271).

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+
+QVariantMap QgsXyzTilesMbtilesAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  const QString outputFile = parameterAsString( parameters, QStringLiteral( "OUTPUT_FILE" ), context );
+
+  QgsProject *project = context.project();
+
+  QgsCoordinateReferenceSystem wgsCrs( "EPSG:4326" );
+  QgsCoordinateReferenceSystem mercatorCrs( "EPSG:3857" );
+  QgsCoordinateTransform src2wgs( project->crs(), wgsCrs, context.transformContext() );
```
Comment (m-kuhn) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:436]: I think that should be moved to `prepare`.

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+      visibleLayers << layer->layer()->id();
+    }
+  }
+
+  QList< QgsMapLayer * > renderLayers = project->layerTreeRoot()->layerOrder();
+  for ( QgsMapLayer *layer : renderLayers )
+  {
+    if ( visibleLayers.contains( layer->id() ) )
+    {
+      mLayers << layer->clone();
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:140]:
```suggestion
      QgsMapLayer* clonedLayer = layer->clone();
      clonedLayer->moveToThread( nullptr );
      mLayers << clonedLayer;
      
```

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+    mMetaTiles += getMetatiles( wgsExtent, z, mMetaTileSize );
+  }
+
+  mTotalTiles = mMetaTiles.size();
+
+  QEventLoop loop;
+  mEventLoop = &loop;
+  startJobs();
+  loop.exec();
+
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:296]:
```suggestion
   qDeleteAll( mLayers );
   mLayers.clear();
```

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+  QgsRectangle wgsExtent = src2wgs.transformBoundingBox( mExtent );
+
+  for ( int z = mMinZoom; z <= mMaxZoom; z++ )
+  {
+    if ( feedback->isCanceled() )
+      break;
+
+    mMetaTiles += getMetatiles( wgsExtent, z, mMetaTileSize );
+  }
+
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:289]:
```suggestion
   for ( QgsMapLayer* layer : std::as_const( mLayers ) )
   {
       layer->moveToThread( QThread::currentThread() );
   }
```

File: src/analysis/processing/qgsalgorithmxyztiles.cpp (hunk 1 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmxyztiles.cpp]: Thanks, done now.

File: src/analysis/processing/qgsalgorithmxyztiles.cpp (hunk 2 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmxyztiles.cpp]: Fixed, thanks.

File: src/analysis/processing/qgsalgorithmxyztiles.cpp (hunk 4 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmxyztiles.cpp:290]: Fixed, thanks.

File: src/analysis/processing/qgsalgorithmxyztiles.cpp (hunk 3 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmxyztiles.cpp:297]: Fixed, thanks.

**Comment by nyalldawson** [comment]
@alexbruy the clang tidy warning can be ignored here, but the failing cppcheck will need to be fixed

**Review by nyalldawson — APPROVED** [review]

**Comment by agiudiceandrea** [comment]
Please see the issue report: https://github.com/qgis/QGIS/issues/55489.

**Comment by agiudiceandrea** [comment]
Please see the issue report: https://github.com/qgis/QGIS/issues/56382.

**Comment by agiudiceandrea** [comment]
Hi @alexbruy please see the issue report: https://github.com/qgis/QGIS/issues/60526.

**Pull request merged**